storage/src/test/java/org/apache/kafka/storage/internals/log/LocalLogTest.java

```diff
@@ -0,0 +1,785 @@
+            return new SimpleRecord(MOCK_TIME.milliseconds(), key.getBytes(), value.getBytes());
+        }
+
+        static KeyValue fromRecord(Record record) {
+            String key = record.hasKey()
+                ? StandardCharsets.UTF_8.decode(record.key()).toString()
+                : "";
+            String value = record.hasValue()
+                ? StandardCharsets.UTF_8.decode(record.value()).toString()
+                : "";
+            return new KeyValue(key, value);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            KeyValue keyValue = (KeyValue) o;
+            return Objects.equals(key, keyValue.key) && Objects.equals(value, keyValue.value);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(key, value);
+        }
+    }
+
+    private List<SimpleRecord> kvsToRecords(List<KeyValue> keyValues) {
+        return keyValues.stream().map(KeyValue::toRecord).collect(Collectors.toList());
+    }
+
+    private List<KeyValue> recordsToKvs(Iterable<Record> records) {
+        List<KeyValue> keyValues = new ArrayList<>();
+        for (Record record : records) {
+            keyValues.add(KeyValue.fromRecord(record));
+        }
+        return keyValues;
+    }
+
+    private void appendRecords(List<SimpleRecord> records, long initialOffset) throws IOException {
+        log.append(initialOffset + records.size() - 1,
+                records.get(0).timestamp(),
+                initialOffset,
+                MemoryRecords.withRecords(initialOffset, Compression.NONE, 0, records.toArray(new SimpleRecord[0])));
+    }
+
+    private FetchDataInfo readRecords(long startOffset) throws IOException {
+        return readRecords(
+                startOffset,
+                log.segments().activeSegment().size(),
+                log.logEndOffsetMetadata()
+        );
+    }
+
+    private FetchDataInfo readRecords(int maxLength) throws IOException {
+        return readRecords(
+                0L,
+                maxLength,
+                log.logEndOffsetMetadata()
+        );
+    }
+
+    private FetchDataInfo readRecords(long startOffset, LogOffsetMetadata maxOffsetMetadata) throws IOException {
+        return readRecords(
+                startOffset,
+                log.segments().activeSegment().size(),
+                maxOffsetMetadata
+        );
+    }
+
+    private FetchDataInfo readRecords(
+                            long startOffset,
+                            int maxLength,
+                            LogOffsetMetadata maxOffsetMetadata) throws IOException {
+        return log.read(startOffset,
+                maxLength,
+                false,
+                maxOffsetMetadata,
+                false);
+    }
+
+    private static List<Record> mkList(Iterable<Record> iterable) {
```

Comment:
this can be replaced by `Utils.toList`

Reply:
Yes, I searched for this method but couldn't find it!